code/povelikin-rostislav/src/main/java/ru/unn/agile/redblacktree/RedBlackTree.java

```diff
@@ -0,0 +1,314 @@
+package ru.unn.agile.redblacktree;
+
+public class RedBlackTree {
+    private class Node {
+        private int key, color = black;
+        private Node left = nil, right = nil, parent = nil;
+
+        Node(final int key) {
+            this.key = key;
+        }
+    }
+
+    private final int red = 0;
+    private final int black = 1;
+    private final Node nil = new Node(-1);
+
+    private Node root = nil;
+    private int size = 0;
+
+    public RedBlackTree() { }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public int getSize() {
+        return size;
+    }
+
+    public boolean remove(final int value) {
+        Node z = getNode(value);
+
+        if (z != nil) {
+            --size;
+        } else {
+            return false;
+        }
+
+        Node x;
+        Node y = z;
+
+        int yOriginalColor = y.color;
+
+        if (z.left == nil) {
+            x = z.right;
+            transplant(z, z.right);
+        } else if (z.right == nil) {
+            x = z.left;
+            transplant(z, z.left);
+        } else {
+            y = treeMinimum(z.right);
+            yOriginalColor = y.color;
+            x = y.right;
+            if (y.parent == z) {
+                x.parent = y;
+            } else {
+                transplant(y, y.right);
+                y.right = z.right;
+                y.right.parent = y;
+            }
+
+            transplant(z, y);
+
+            y.left = z.left;
+            y.left.parent = y;
+            y.color = z.color;
+        }
+
+        if (yOriginalColor == black) {
+            deleteFix(x);
+        }
+
+        return true;
+    }
+
+    public boolean find(final int expected) {
+        Node found = getNode(expected);
+        return found != nil;
+    }
+
+    private Node getNode(final int expected) {
+        if (root == nil) {
+            return nil;
+        }
+
+        Node temp = root;
+
+        while (true) {
+            if (expected == temp.key) {
+                return temp;
+            } else if (expected < temp.key) {
+                if (temp.left == nil) {
+                    return nil;
+                } else {
+                    temp = temp.left;
+                }
+            } else {
+                if (temp.right == nil) {
+                    return nil;
+                } else {
+                    temp = temp.right;
+                }
+            }
+        }
+    }
+
+    public void insert(final int value) {
+        ++size;
```

Comment:
I think better to perform this operation after all actions, when we sure that new element really added. Same for removing.

Reply:
Refactored